### PR TITLE
Add Pagination component to @uifabric/experiments

### LINKED ISSUE #349: New Component: Pagination
Hi all,

I would like to implement react fabric pagination component but actually we haven't. Could you implement this control ?

It would be great if we could have this option directly on the list component or just the HTML tag with CSS class in order to integrate in our custom component.

Thank you

### PR DESCRIPTION
**Pull request by caradong**
#### Pull request checklist
- [x] Addresses an existing issue: Fixes #349 
- [x] Include a change request file using `$ npm run change`

#### Description of changes
initial checkin of pagination control into experiment package

pagination1113

#### Focus areas to test

pagination

### CODE REVIEW

File: packages/experiments/src/components/Pagination/PageNumber.tsx [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    super(props);
+  }
+
+  public render(): JSX.Element {
+    const { className, ariaLabel, page, selected } = this.props;
+
+    return (
+      <button
+        key={page}
+        className={className}
```
Comment (kenotron) [inline packages/experiments/src/components/Pagination/PageNumber.tsx]: Thinking that you can use `getNativeProps` here to wrap a lot of existing button props.

File: packages/experiments/src/components/Pagination/PageNumber.tsx [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        aria-selected={selected}
+        aria-label={ariaLabel}
+      >
+        {page}
+      </button>
+    );
+  }
+
+  private onClick = () => {
+    this.props.applyPage(this.props.page - 1);
```
Comment (kenotron) [inline packages/experiments/src/components/Pagination/PageNumber.tsx]: I don't think we should do the page number to index conversion for consumers of this control. If the `props.page` expected a 1-based numbering system, then I think the operations should expect the same 1-based number

File: packages/experiments/src/components/Pagination/Pagination.base.tsx [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+          <IconButton iconProps={lastPageIconProps} onClick={this.handleLastPage} disabled={!canLast} aria-label={lastPageAriaLabel} />
+        </div>
+      );
+    }
+
+    let visibleItemLabel;
+    if (itemsPerPage && totalItemCount) {
+      const leftItemIndex = selectedPageIndex! * itemsPerPage + 1;
+      const rightItemsIndex = Math.min((selectedPageIndex! + 1) * itemsPerPage, totalItemCount);
+      visibleItemLabel = `${leftItemIndex} - ${rightItemsIndex} of ${totalItemCount}`;
```
Comment (kenotron) [inline packages/experiments/src/components/Pagination/Pagination.base.tsx]: This embeds English text. Probably best to be a function that you can override as a prop?

File: packages/experiments/src/components/Pagination/Pagination.base.tsx [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+  };
+
+  private handleNextPage = () => {
+    this.handleSelectedPage(this.props.selectedPageIndex! + 1);
+  };
+
+  private _pageElement(index: number): JSX.Element {
+    const { pageAriaLabel, pageCount, selectedPageIndex, selectedAriaLabel } = this.props;
+    const isSelected = index === selectedPageIndex;
+    let ariaLabel = pageAriaLabel && `${pageAriaLabel} ${index + 1} of ${pageCount}`;
```
Comment (kenotron) [inline packages/experiments/src/components/Pagination/Pagination.base.tsx]: again, careful that you're embedding english here

File: packages/experiments/src/components/Pagination/Pagination.base.tsx (hunk 3 above)
Comment (caradong) [reply packages/experiments/src/components/Pagination/Pagination.base.tsx]: I was thinking to ask someone from you team how to do this. Essentially, this is a string template in the format of {0} - {1} of {2} . Is there any existing component in fabric that take a string template as property for either display or aria label ? If there is any existing best practice I can follow, that would be very helpful 

---
In reply to: [233571920](https://github.com/OfficeDev/office-ui-fabric-react/pull/7100#discussion_r233571920) [](ancestors = 233571920)

File: packages/experiments/src/components/Pagination/PageNumber.tsx (hunk 1 above)
Comment (caradong) [reply packages/experiments/src/components/Pagination/PageNumber.tsx]: I tried it and look like the only property I was able to get rid of is the className. I am not very familiar with getNavtiveProps, please let me know if I am not using it in the correct way 

---
In reply to: [233569584](https://github.com/OfficeDev/office-ui-fabric-react/pull/7100#discussion_r233569584) [](ancestors = 233569584)

File: packages/experiments/src/components/Pagination/Pagination.base.tsx (hunk 4 above)
Comment (caradong) [reply packages/experiments/src/components/Pagination/Pagination.base.tsx]: please see my other comment

---
In reply to: [233575229](https://github.com/OfficeDev/office-ui-fabric-react/pull/7100#discussion_r233575229) [](ancestors = 233575229)

**Comment by caradong** [comment]
@kenotron I talked to @KevinTCoughlin and @micahgodbolt on Teams and followed their suggestion about the localization needs of strings.

File: packages/experiments/src/components/Pagination/Pagination.types.ts [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+  buttons = 0,
+
+  /**
+   * Display Pagination as Combo Box
+   */
+  comboBox = 1
+}
+
+export interface IPaginationString {
+  /** the divider between the 'from item index' and the 'to item index', e.g. '-' */
```
Comment (micahgodbolt) [inline packages/experiments/src/components/Pagination/Pagination.types.ts]: can you do single line comments like this
```
 /**
   * The string object
   */
```

File: packages/experiments/src/components/Pagination/Pagination.types.ts [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+  /**
+   * provide a string for the visible item label if localization of this string is needed.
+   * @param fromItemIndex calculated based on @see totalItemCount , @see itemsPerPage and @see selectedPageIndex
+   * @param toItemIndex calculated based on @see totalItemCount , @see itemsPerPage and @see selectedPageIndex
+   * @param totalItem @see totalItemCount
+   * @param divider @see the divider string between @param fromItemIndex and @param toItemIndex
+   * @param stringOf the string 'of'
+   * @default default string format: `${fromItemIndex} ${divider} ${toItemIndex} ${stringOf} ${totalItemCount}`
+   */
+  onRenderVisibleItemLabel?: (
```
Comment (micahgodbolt) [inline packages/experiments/src/components/Pagination/Pagination.types.ts]: can we just pass props through here? For onRenderX you can use IRenderFunction which is props + default render.

File: packages/experiments/src/components/Pagination/Pagination.types.ts (hunk 5 above)
Comment (caradong) [reply packages/experiments/src/components/Pagination/Pagination.types.ts]: sure.

---
In reply to: [235120457](https://github.com/OfficeDev/office-ui-fabric-react/pull/7100#discussion_r235120457) [](ancestors = 235120457)

File: packages/experiments/src/components/Pagination/Pagination.types.ts (hunk 6 above)
Comment (caradong) [reply packages/experiments/src/components/Pagination/Pagination.types.ts]: updated.

File: packages/experiments/src/components/Pagination/PageNumber.tsx [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+export class PageNumber extends React.Component<IPageNumberProps, {}> {
+  constructor(props: IPageNumberProps) {
+    super(props);
+  }
+
+  public render(): JSX.Element {
+    const { ariaLabel, page, selected } = this.props;
+
+    return (
+      <button
```
Comment (micahgodbolt) [inline packages/experiments/src/components/Pagination/PageNumber.tsx]: Is it possible to use Fabric buttons here?

File: packages/experiments/src/components/Pagination/PageNumber.types.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
+export interface IPageNumberProps {
+  /** The call back function when click on the page number . */
+  applyPage: (page: number) => void;
```
[the remark this answers is not in the document]
Comment (caradong) [reply packages/experiments/src/components/Pagination/PageNumber.types.ts]: updated to be 'onClick'

---
In reply to: [235191795](https://github.com/OfficeDev/office-ui-fabric-react/pull/7100#discussion_r235191795) [](ancestors = 235191795)

**Comment by micahgodbolt** [comment]
What you have now is great. The question is where we plan to go in the future and how that might affect the way this component is customized. There are not changes needed, I just wanted to point out the possibility of these improvements in the future.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by betrue-final-final** [comment]
FYI, the numbered links versions are designed for specific cases that Cara's team would like to use them for, but are confusing for screen reader users as the numbers change and it's not completely clear how to get to the pages not shown.

**Comment by caradong** [comment]
@micahgodbolt @kenotron I have addressed all the comments. Let me know if there is any further feedback. Thanks

**Comment by micahgodbolt** [comment]
Sorry about the delay @caradong. I'm happy with this as is. Could you create a new issue for pagination to collect the things we want to do before this comes out of experiments, and link this PR to that issue.

Customizing links is one thing that came to mind.

**Comment by KevinTCoughlin** [comment]
> 
> 
> Sorry about the delay @caradong. I'm happy with this as is. Could you create a new issue for pagination to collect the things we want to do before this comes out of experiments, and link this PR to that issue.
> 
> Customizing links is one thing that came to mind.

@caradong can you please do the above and then I think this can be merged into experiments.

**Comment by caradong** [comment]
@KevinTCoughlin thank you Kevin. I created issue https://github.com/OfficeDev/office-ui-fabric-react/issues/7254 and feel free to add more work there.

**Comment by KevinTCoughlin** [comment]
> 
> 
> @KevinTCoughlin thank you Kevin. I created issue #7254 and feel free to add more work there.

Cool, added some feedback during my last test of the control.

**Pull request merged**